File: node/exercises/scrab-bag.js
const colors = require('colors')
const quitCommand = require('../helpers/quitCommand')
const scrabbleLetters = require('../constants/scrabbleLetters')

const letters = Object.keys(scrabbleLetters)
let scrabBag = Object.assign(scrabbleLetters)

console.log('Enter the tiles that have already been played.'.green)

process.stdin.resume();
process.stdin.setEncoding('utf8');
process.stdin.on('data', function(input){
  if(input.indexOf('/') === 0) {
    quitCommand(input, process)
  } else {
    input = input.replace(/\W|\d/g, '').toUpperCase().split('')
    console.log(input)

    input.forEach(letter => {
      scrabBag[letter].count = scrabBag[letter].count - 1
    })

    let remainingLetters = {}

    letters.forEach(letter => {
      let count = scrabBag[letter].count;
      if(remainingLetters.hasOwnProperty(count)){
        remainingLetters[count].push(letter)
      } else {
        remainingLetters[count] = [letter]
      }
    })

    let remainingLetterKeys = Object.keys(remainingLetters)

    console.log('Letters remaining in the bag:'.green)
    remainingLetterKeys.forEach(key => {
      let currentLetters = []
      console.log(colors.green(key + ': ' + remainingLetters[key].join(', ')))
    })
    process.disconnect()
  }
})

File: ruby/exercises/scrab_bag.rb
require 'colorize'
require_relative('../constants/scrabble_letters')
require_relative('../helpers/exit_prompt')
require_relative('../helpers/quit_to_menu')

class ScrabBag
  def initialize
    @scrab_bag = $scrabble_letters.clone
    @remaining_letters = {}
    @letters = @scrab_bag.keys
    start!
  end

  def start!
    puts('Enter the letters that have already been played.'.yellow)
    input = gets
    quit_to_menu(input)
    input.gsub!(/\W|\d/, '')
    input = input.upcase.split('')

    input.each { |letter| reduce_letter_count(letter) }
    @letters.each { |letter| get_remaining_letters(letter) }

    puts('Letters remaining in the bag:'.green)
    @remaining_letters.keys.sort.each { |key|
      puts(key.to_s.green + ': '.green + @remaining_letters[key].join(', ').yellow)
    }
    display_exit_prompt
  end

  def reduce_letter_count(letter)
    @scrab_bag[letter.to_sym][:count] = @scrab_bag[letter.to_sym][:count] -1
  end

  def get_remaining_letters(letter)
    count = @scrab_bag[letter.to_sym][:count]
    if(@remaining_letters.has_key?(count))
      @remaining_letters[count].push(letter)
    else
      @remaining_letters[count] = [letter]
    end
  end

  def display_exit_prompt
    exit_prompt
  end

  private :reduce_letter_count, :get_remaining_letters, :display_exit_prompt
end

ScrabBag.new
